src/coreclr/jit/fginline.cpp

```diff
@@ -586,8 +604,55 @@ class SubstitutePlaceholdersAndDevirtualizeWalker : public GenTreeVisitor<Substi
                 }
 
                 CORINFO_CONTEXT_HANDLE contextInput = context;
+                context                             = nullptr;
                 m_compiler->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &contextInput, &context,
                                                 isLateDevirtualization, explicitTailCall);
+                if (context != nullptr)
+                {
+#if defined(DEBUG)
+                    IL_OFFSET ilOffset = call->gtRawILOffset;
+#else
+                    IL_OFFSET ilOffset = m_curStmt->GetDebugInfo().GetLocation().GetOffset();
+#endif // defined(DEBUG)
+                    CORINFO_CALL_INFO callInfo = {};
+                    callInfo.hMethod           = method;
+                    callInfo.methodFlags       = methodFlags;
+                    m_compiler->impMarkInlineCandidate(call, context, false, &callInfo, ilOffset);
+
+                    if (call->IsInlineCandidate())
+                    {
+                        CallArg* retBuffer = call->gtArgs.GetRetBufferArg();
+
+                        // If the call is the top-level expression in a statement, and it returns void,
+                        // there will be no use of its return value, and we can just inline it directly.
+                        // In this case we don't need to create a RET_EXPR node for it.
+                        if (parent != nullptr || genActualType(call->TypeGet()) != TYP_VOID || retBuffer != nullptr)
+                        {
+                            Statement* stmt = m_compiler->gtNewStmt(call);
+                            m_compiler->fgInsertStmtBefore(m_compiler->compCurBB, m_curStmt, stmt);
+                            GenTreeRetExpr* retExpr =
+                                m_compiler->gtNewInlineCandidateReturnExpr(call->AsCall(),
+                                                                           genActualType(call->TypeGet()));
+
+                            call->GetSingleInlineCandidateInfo()->retExpr = retExpr;
+
+                            m_curStmt = stmt;
+                            *pTree    = retExpr;
+                        }
+
+                        if (retBuffer != nullptr)
+                        {
+                            call->GetSingleInlineCandidateInfo()->preexistingSpillTemp =
+                                retBuffer->GetNode()->AsLclVarCommon()->GetLclNum();
```

Comment:
What guarantees that the return buffer is of the correct shape?

Reply:
It turns out that `preexistingSpillTemp` here is not used at all. Resolved by removing it.